### PR TITLE
Sanitize world names when creating them, and store the original name elsewhere

### PR DESCRIPTION
**Pull request by Df458**
This PR addresses #4875 by preventing potentially invalid paths from being created when making a new world.

The way it works is fairly simplistic: If a character matches against a blacklist (containing path characters and chars forbidden by certain filesystems), it is replaced by '\_' in the path. The display name is stored in `world.mt`, and duplicate file names are resolved by adding an incrementing suffix (\_1, \_2, \_3, etc) as suggested in the original issue. Additionally, the world_ prefix is added if the name matches a blacklist of 'reserved names' that some filesystems have trouble with (see the original issue for more details).

## Testing
Create worlds, and put garbage into their names. Try making worlds with different names but duplicate sanitized names. Try hosting/launching these worlds. Ensure that these worlds aren't being duplicated disappearing for any reason. Check anywhere in the UI that you would expect to see the world name: Is the name you entered there, or is it the sanitized version?

## Known Issues
- Names are still case-sensitive. On case-insensitive filesystems, collisions can probably still happen (needs testing)

### CODE REVIEW

**Comment by appgurueu** [comment]
> the world is not created (normal) but no feedback is given to the user about this
I would go with a plain error message saying "World could not be created: too many worlds with similar names, chose a different one."

**Comment by Df458** [comment]
Rebased on 5.2. I'll try to finish this feature up this week so that it can be reviewed/merged.

**Comment by Df458** [comment]
So I've added the necessary code to tell the user more directly when a world can't be created due to hitting the limit--but in retrospect, I'm starting to wonder if an arbitrary limit is even really necessary.

I mostly created it to eventually break out of the while loop, but aside from slowness (and difficulty manually managing the directories) there's no other real downside to creating similar worlds. We could just raise the limit to max int, which I seriously doubt anyone will manage to hit.

Any thoughts on this

**Comment by paramat** [comment]
For me it seems better and simpler to just refuse a non-compliant world name in main menu and display a message of explanation. No need to store the display name, avoids duplicates, as well as teaching users how to properly name their worlds and so avoid repeating the problem.

**Comment by Df458** [comment]
> For me it seems better and simpler to just refuse a non-compliant world name in main menu and display a message of explanation.

So what's a non-compliant world name? Obviously, we should block certain characters like '/'. But on case-insensitive filesystems, I suspect we already have weirdness with lowercase/uppercase world names (which I'm pretty sure minetest doesn't check for). Do we also update our duplicate check to be case-insensitive, although some systems support them? What about special characters that aren't path separators, but are still barred from windows while being allowed on linux? Do we also explicitly check for 'NUL' while we're at it? How about on non-NTFS filesystems?

...and if the answer to all that is yes, how exactly do you propose we explain the resulting ruleset to a user? I'm not saying it's impossible, but I wouldn't call it 'simpler'.

**Comment by Df458** [comment]
Bump. What do other contributors and coredevs think? This PR has been open for close to a month and @paramat disapproves of the current approach, so if I don't see any support for this in the next week or so I'll close it to free up the backlog.

**Comment by sfan5** [comment]
Sounds okay to me in principle but Unicode is definitely a problem that has to be considered.
Perhaps it would be possible to turn just the characters that cause problems into `_` (instead of everything outisde `A-Za-z0-9_-`).

**Comment by rubenwardy** [comment]
> ...and if the answer to all that is yes, how exactly do you propose we explain the resulting ruleset to a user? I'm not saying it's impossible, but I wouldn't call it 'simpler'.

You don't. You save worlds like `nul` as `nul1` or some other non-problematic name

**Comment by Df458** [comment]
> Perhaps it would be possible to turn just the characters that cause problems into _

I'm a little wary of this because I suspect we'll miss something along the way, but even an incomplete blacklist is better than no checks at all. I'll change the PR to block path-related characters and the list [here](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file?redirectedfrom=MSDN#file-and-directory-names) for starters.


 > Minetest should probably avoid duplicate world display names though, but duplicate world folder names should need to be mitigated

Pretty sure this PR still blocks duplicate display names, since there's a separate check for that. The only time where this could happen is if a user manually changes the display name in a text editor, or downloads a world with a duplicate name but non-duplicate directory name. I think those edge-cases are fine to leave, since they should be pretty uncommon.

I still think we should keep the `world_` prefix as-is, since it prevents cases like `nul` without adding a bunch of special checks into the code.

Does this sound acceptable?

**Comment by rubenwardy** [comment]
That sounds good to me

**Comment by sfan5** [comment]
> I still think we should keep the `world_` prefix as-is, since it prevents cases like `nul` without adding a bunch of special checks into the code.

I often type out the filesystem paths to my worlds and would not like to be inconvenienced by Microsoft's implementation choices when I'm not even using Windows.
So my preference would be to keep an explicit blacklist of the (`con`, `nul`, ...) filenames for Windows. I don't mind if `:` turns to `_` even on Linux, just the `world_` prefix is what annoys me.

**Comment by Df458** [comment]
I've replaced the whitelist with a blacklist as requested. I also ran some tests and modified the function to handle Unicode characters better. I'm not well-versed in handling Unicode in C/C++, so please let me know if I've done anything monumentally stupid by accident.

> So my preference would be to keep an explicit blacklist of the (con, nul, ...) filenames for Windows.

What do you propose we do when we encounter a blacklisted name? Should we add the `world_` prefix then, or do something else?

**Comment by Df458** [comment]
One more thing: I'd like someone on Windows to test what happens when making two worlds with the same letters but different casing (ie 'test' and 'TEST'). Please let us know what happens, I have a hunch but I'd prefer actual data for this case.

**Comment by sfan5** [comment]
> What do you propose we do when we encounter a blacklisted name? Should we add the world_ prefix then, or do something else?

Considering it's better when it "just works", adding the `world_` prefix is a better approach.

**Comment by Df458** [comment]
@sfan5 / @rubenwardy -- I believe I've implemented what was asked of me. If you've got the time, can you take a look and confirm if things are working as you'd like them to now?

File: src/script/lua_api/l_mainmenu.cpp [hunk 1]
```diff
@@ -618,16 +618,17 @@ int ModApiMainMenu::l_create_world(lua_State *L)
...
 
 	if ((gameidx >= 0) &&
 			(gameidx < (int) games.size())) {
 
 		// Create world if it doesn't exist
-		if (!loadGameConfAndInitWorld(path, games[gameidx])) {
-			lua_pushstring(L, "Failed to initialize world");
+		std::string error;
+		if (!loadGameConfAndInitWorld(path, name, games[gameidx], true, &error)) {
+			lua_pushstring(L, std::string("Failed to initialize world: " + error).c_str());
```
Comment (rubenwardy) [inline src/script/lua_api/l_mainmenu.cpp:631]: the string construction here is redundant, you probably meant 

```c++
(std::string("Failed to initialize world: ") + error).c_str()
```

which can be simplified to

```cpp
("Failed to initialize world: "s + error).c_str()
```

But given the code in your PR compiles, this is somehow OK:

```cpp
("Failed to initialize world: " + error).c_str()
```

File: src/server.cpp [hunk 2]
```diff
@@ -360,8 +360,11 @@ void Server::init()
 	infostream << "- game:   " << m_gamespec.path << std::endl;
 
 	// Create world if it doesn't exist
-	if (!loadGameConfAndInitWorld(m_path_world, m_gamespec))
-		throw ServerError("Failed to initialize world");
+	std::string error;
+	if (!loadGameConfAndInitWorld(m_path_world,
+				fs::GetFilenameFromPath(m_path_world.c_str()),
```
Comment (rubenwardy) [inline src/server.cpp:365]: indent too far

File: src/util/string.cpp [hunk 3]
```diff
@@ -889,3 +889,91 @@ std::wstring translate_string(const std::wstring &s)
 	return translate_string(s, g_client_translations);
 #endif
 }
+
+static const std::wstring disallowed_dir_names[] {
```
Comment (rubenwardy) [inline src/util/string.cpp:893]: consider making this a `std::array` so the length becomes a compile error if it does not match

File: src/util/string.cpp [hunk 4]
```diff
@@ -889,3 +889,91 @@ std::wstring translate_string(const std::wstring &s)
...
+	L"LPT7",
+	L"LPT8",
+	L"LPT9",
+};
+static const int disallowed_name_count = 22;
+
+/**
+ * List of characters that are blacklisted from created directories
+ */
+static const wchar_t disallowed_path_chars[] {
```
Comment (rubenwardy) [inline src/util/string.cpp:924]: same here

**Comment by rubenwardy** [comment]
If you copy and rename a world, the name stays the same in the world list as _all_ worlds are given `world_name`s. I don't necessarily think this is a problem

**Comment by paramat** [comment]
Df458, i only had a partial understanding of the situation, so i retract my earlier suggestion.
I did not 'disapprove' of the implementation as in 'officially opposes the PR', i did not feel that strongly about it.

File: src/script/lua_api/l_mainmenu.cpp (hunk 1 above)
Comment (Df458) [reply src/script/lua_api/l_mainmenu.cpp]: Fixed with the first option. The `""s` notation doesn't compile on my end for whatever reason, but you're correct that I missed those parens.

**Comment by Df458** [comment]
Rebased on 5.4.0

By the way, I don't think the "Action / change needed" label applies anymore. All the review comments were resolved a couple weeks ago.

File: src/util/string.cpp [hunk 5]
```diff
@@ -889,3 +890,89 @@ std::wstring translate_string(const std::wstring &s)
...
+	L"LPT6",
+	L"LPT7",
+	L"LPT8",
+	L"LPT9",
+};
+
+/**
+ * List of characters that are blacklisted from created directories
+ */
+static const std::array<wchar_t, 10> disallowed_path_chars = {
```
Comment (SmallJoker) [inline src/util/string.cpp:924]: If preferred: This could be a simple string, without manual iteration:
```C++
static const std::wstring disallowed_path_chars = L"<>:\"/\\|?*.";

is_valid = tmp.find_first_of(disallowed_path_chars) == std::wstring::npos;
```

File: src/util/string.cpp [hunk 6]
```diff
@@ -889,3 +890,89 @@ std::wstring translate_string(const std::wstring &s)
...
+	'.',
+};
+
+/**
+ * Sanitize the name of a new directory. This consists of two stages:
+ * 1. Check for 'reserved filenames' that can't be used on some filesystems
+ *	and add a prefix to them
+ * 2. Remove 'unsafe' characters from the name by replacing them with '_'
+ */
+std::string sanitizeDirName(const std::string &str, const std::string &unsafe_prefix)
```
Comment (SmallJoker) [inline src/util/string.cpp:947]: `unsafe_prefix` .. the prefix is unsafe?
What would you think about `optional_prefix` or so?

File: src/util/string.cpp [hunk 7]
```diff
@@ -889,3 +890,89 @@ std::wstring translate_string(const std::wstring &s)
...
+{
+	std::wstring tmp = utf8_to_wide(str);
```
Comment (SmallJoker) [inline src/util/string.cpp:949]:
```suggestion
	std::wstring safe_name = utf8_to_wide(str);
```
..or something like that.

File: src/util/string.cpp [hunk 8]
```diff
@@ -889,3 +890,89 @@ std::wstring translate_string(const std::wstring &s)
...
+
+	for(std::wstring disallowed_name : disallowed_dir_names) {
+		if (str_equal(tmp, disallowed_name, true)) {
+			tmp = utf8_to_wide(unsafe_prefix) + tmp;
```
Comment (SmallJoker) [inline src/util/string.cpp:953]: `break;` below.
Missing space after `for`

File: src/server.cpp [hunk 9]
```diff
@@ -356,8 +356,11 @@ void Server::init()
 	infostream << "- game:   " << m_gamespec.path << std::endl;
 
 	// Create world if it doesn't exist
-	if (!loadGameConfAndInitWorld(m_path_world, m_gamespec))
-		throw ServerError("Failed to initialize world");
+	std::string error;
+	if (!loadGameConfAndInitWorld(m_path_world,
+            fs::GetFilenameFromPath(m_path_world.c_str()),
+            m_gamespec, false, &error))
```
Comment (SmallJoker) [inline src/server.cpp:362]: Spaces -> tabs

File: src/content/subgames.cpp [hunk 10]
```diff
@@ -288,8 +306,27 @@ std::vector<WorldSpec> getAvailableWorlds()
...
+	if (mustCreateNew) {
+		int counter = 1;
+		while (fs::PathExists(final_path) && counter < MAX_WORLD_NAMES) {
+			final_path = path + "_" + std::to_string(counter);
+			counter++;
+		}
+
+		if (fs::PathExists(final_path)) {
+			if (error != nullptr) {
+				*error = "Too many similar filenames";
```
Comment (SmallJoker) [inline src/content/subgames.cpp:324]: Wouldn't this be an ideal use-case for exceptions?

File: src/content/subgames.cpp [hunk 11]
```diff
@@ -288,8 +306,27 @@ std::vector<WorldSpec> getAvailableWorlds()
...
 }
 
-bool loadGameConfAndInitWorld(const std::string &path, const SubgameSpec &gamespec)
+bool loadGameConfAndInitWorld(const std::string &path, const std::string &name,
+		const SubgameSpec &gamespec, bool mustCreateNew, std::string *error)
 {
+	std::string final_path = path;
+
+	// If we're creating a new world, ensure that the path isn't already taken
+	if (mustCreateNew) {
```
Comment (SmallJoker) [inline src/content/subgames.cpp:315]: Low priority: According to the guidelines, local variables and arguments should use underscores.

Suggestion:
```suggestion
	if (create_new_world) {
```

File: src/content/subgames.cpp (hunk 10 above)
Comment (Df458) [reply src/content/subgames.cpp:324]: I'm always a little wary of throwing exceptions in a gamedev context, but since this isn't a hotspot it should be fine. Do you know any existing exceptions that would fit the bill here, or would you prefer that I make a new one?

Comment (SmallJoker) [reply src/content/subgames.cpp:324]: A generic exception (`BaseException`) likely suffices here. Yet I'll leave the decision whether to use exceptions to you.
Oh and `InvalidFilenameException` is currently unused and could be removed...

File: src/util/string.cpp (hunk 8 above)
Comment (SmallJoker) [reply src/util/string.cpp]: Fixed this.

File: src/content/subgames.cpp (hunk 10 above)
Comment (Df458) [reply src/content/subgames.cpp:324]: I thought this was a good suggestion at first, but after writing most of the necessary code I found that it makes the error checking in `l_mainmenu.cpp` kinda messy as a result... we could make it work by making this function return void and throwing an exception for the other failure state as well, but I'd like your opinion on that first.

Comment (SmallJoker) [reply src/content/subgames.cpp:324]: It would be simple as
```C++
void loadGameConfAndInitWorld()
{
	throw BaseException("Some error");
}

try {
	loadGameConfAndInitWorld(path, name, games[gameidx], true))
	lua_pushnil(L);
} catch (BaseException &e) {
	lua_pushstring(L, ("Failed to initialize world: " + e.what()).c_str());
}
```

Comment (Df458) [reply src/content/subgames.cpp:324]: Gotcha, I'll go ahead and replace both then :)

**Comment by SmallJoker** [comment]
:+1: works still as expected.
